File: code/exercise_solutions/Workshop 1 - data management.R
library(tidyverse)
library(slider)

## -----------------------------------------------------------------------------
## Funky functions
## -----------------------------------------------------------------------------
func_nyears = function(x, .f, date, interval, ...){
  # Uses the function .f on all observations within (interval) years prior to
  # each individual observation in x
  res = rep(0,length(x))
  ints = interval(date, date)
  
  for(i in 1:length(x)){
    int_xi = interval(date[i] %m-% years(interval), date[i])
    index = int_overlaps(int_xi, ints)
    res[i] = .f(x[index], ...)
  }
  return(res)
}


## -----------------------------------------------------------------------------
## Load data
## -----------------------------------------------------------------------------

# Data should be saved in a folder named "Data" in the working directory
n_years <- 1

baseline <- read.csv("data/baseline_data.csv")
diag     <- read.csv("data/diag_data.csv") %>% select(-X)
dict     <- read.csv("data/dict_data.csv") %>% select(-X)
quest    <- read.csv("data/quest_data.csv") %>% select(-X)

blood <- 
  read.csv("data/blood_data.csv") %>% 
  select(-X) %>% 
  rename(id = ..record.id)

treat <- 
  read.csv("data/treat_data.csv") %>% 
  select(-X) %>% 
  rename(id = record_id)

visits <- 
  read.csv("data/visit_date.csv") %>% 
  select(id, date = visit_date) %>% 
  mutate(visit = T)

events <- read.csv("data/events.csv") %>% 
  select(-X)

## -----------------------------------------------------------------------------
## Translate diagnosis codes
## -----------------------------------------------------------------------------

diag <- 
  diag %>% 
  left_join(dict, by = "code") %>% 
  select(-code)


## -----------------------------------------------------------------------------
## Removing variable with >20% missing (for a more rigorous approach
## this should be evaluated based on training data only, but here for simplicity 
## done on all available data)
## -----------------------------------------------------------------------------

baseline <- 
  baseline %>% 
  select_if(~ mean(is.na(.)) < 0.2)

blood <- 
  blood %>% 
  select_if(~ mean(is.na(.)) < 0.2)

quest <- 
  quest %>% 
  select_if(~ mean(is.na(.)) < 0.2)

treat <- 
  treat %>% 
  select_if(~ mean(is.na(.)) < 0.2) %>%
  fill(aspirin:statins, .direction = "down") %>% 
  mutate(across(aspirin:statins, function(x){x[is.na(x)] <- mean(x, na.rm = T); return(x)}))
# Note that NA's are filled in for the treat dataset already here
# This is useful when done before combining the data set


## -----------------------------------------------------------------------------
## Training, validation, calibration, and test set
## -----------------------------------------------------------------------------

set.seed(1306)

idx <- 
  sample(c("train", "validation", "calibration", "test"), 
         size = nrow(baseline), replace  = T, prob = c(0.8, 0.05, 0.05, 0.1)) 

splits <- tibble(id = baseline$id, split = idx) 

# OBS: here we have approximately an equal number of visits (=prediction 
# time points) per patient, so just splitting data on a patient-level. 


## -----------------------------------------------------------------------------
## One dataset with a row per time stamp to contain all observed features
## -----------------------------------------------------------------------------

timevar <- 
  # Merge data - one row per time stamp
  full_join(blood, diag, multiple = "all", by = c("id", "sample_date")) %>% 
  full_join(quest, multiple = "all", by = c("id", "sample_date" = "date.x")) %>% 
  full_join(treat, multiple = "all", by = c("id", "sample_date" = "treat_start_date")) %>% 
  full_join(visits, multiple = "all", by = c("id", "sample_date" = "date")) %>% 
  full_join(baseline, multiple = "all", by = "id") %>% 
  full_join(events %>% filter(type %in% c("minor","other","major")) %>% select(-"type"),
            multiple = "all", by = c("id","sample_date" = "date")) %>%
  rename(date = sample_date) %>% 
  left_join(splits, by = "id") %>%
  mutate(date = as.POSIXlt(date, tz = "UTC"),
         age = interval(d.birth, date) / years(1)) %>% 
  relocate(age, .after = statins) %>% 
  arrange(id,date)

# OBS: the date is made into a POSIXlt class, as these work well with intervals
# Very useful when creating dynamic variables (using the funky function)


## -----------------------------------------------------------------------------
## Mean values for imputation and quantile values for capping 
## -----------------------------------------------------------------------------

summaries <- 
  timevar %>% 
  filter(split == "train") %>% 
  select(-diag) %>% 
  summarise(across(systolic:age, 
                   ~ mean(., na.rm = T), 
                   .names = "mean_{.col}"),
            across(c(systolic, diastolic, hemoglobin,
                     platelets, creatine, anxiety:statins), 
                   ~ round(mean(., na.rm = T)), 
                   .names = "mean_{.col}"),
            across(systolic:age, 
                   ~ quantile(., p = 0.01, na.rm = T), 
                   .names = "q1_{.col}"), 
            across(systolic:age, 
                   ~ quantile(., p = 0.99, na.rm = T), 
                   .names = "q99_{.col}"))

# Note: q1 and q99 will be equal to min and max in the training set after
# capping


## -----------------------------------------------------------------------------
## Feature engineering
## -----------------------------------------------------------------------------

final_data <-
  timevar %>% 
  # Prediction time points and encoding of sex and smoking
  mutate(visit          = if_else(is.na(visit), F, visit), 
         male           = if_else(sex == "male", 1, 0), 
         smoker_current = if_else(smoking == "current", 1, 0), 
         smoker_former  = if_else(smoking == "former", 1, 0)) %>% 
  # One-hot encoding of diagnoses and events - using carry-forward approach
  group_by(id) %>% 
  mutate(diabetes          = case_when(diag  == "diabetes" ~ 1, T ~ 0), 
         hyperlipidemia    = case_when(diag  == "hyperlipidemia" ~ 1, T ~ 0), 
         hypertension      = case_when(diag  == "hypertension" ~ 1, T ~ 0),
         across(c(diabetes, hyperlipidemia, hypertension), cumsum),
         revascularization = case_when(event == "revascularization" ~ 1, T ~ 0),
         malignancy        = case_when(event == "malignancy" ~ 1, T ~ 0),
         stroke            = case_when(event == "stroke" ~ 1, T ~ 0),
         amputation        = case_when(event == "amputation" ~ 1, T ~ 0),
         infarction        = case_when(event == "infarction" ~ 1, T ~ 0)) %>% 
  # Calculating some summary variables for the cumulated (countable) variables
  mutate(across(revascularization:infarction, cumsum, .names = "n_{.col}"),
         across(revascularization:infarction, ~ func_nyears(., sum, date, n_years),
                .names = "nyear_{.col}"),
         n_treatments     = case_when(is.na(clopridrogel) ~ 0, # this ensures 0s at times where no treatment observations are found
                                      T ~ clopridrogel + ezetimibe + anticlot + aspirin + statins),
         nyear_treatments = func_nyears(n_treatments, sum, date, n_years),
         n_treatments     = cumsum(n_treatments)) %>% 
  select(-diag, -d.birth, -sex, -smoking, -event, -revascularization, -malignancy, -stroke, -amputation, -infarction) %>% 
  # Carry-forward approach for blood, treatment, and questionnaire data: 
  fill(everything(), .direction = "down") %>% 
  # If no value to carry forward, use mean from training data
  replace_na(as.list(summaries %>% 
                       select(starts_with("mean")) %>% 
                       rename_with(~str_replace(., "mean_", "")))) %>% 
  # Create summary variable on the blood dataset in the form x - mean(x) for each person
  mutate(across(systolic:cvscore, ~ slide_vec(., 
                                              .f = function(x){
                                                if(length(x) == 1){return(0)};
                                                tail(x,1) - mean(head(x,-1))},
                                              .before = Inf),
                .names = "diff_{.col}")) %>% 
  # Prediction time points only
  filter(visit) %>% 
  select(-visit) %>% 
  # Difference in the questionnaire answers from visit to visit (summary variable)
  mutate(across(anxiety:vegetable, ~ c(0, diff(.)), .names = "diff_{.col}")) %>% 
  # Outcome: n-year mortality with n defined at the start of the document
  left_join(events %>%
              filter(type == "death") %>% 
              slice_min(order_by = date, by = "id", with_ties = F) %>% 
              select(date, id) %>% 
              rename(d.event = date), by = "id") %>% 
  ungroup() %>% 
  # Create the response variable
  mutate(mevent_nyear = case_when(interval(date, d.event) / years(1) <= n_years ~ 1, 
                                  T ~ 0)) %>% 
  select(-d.event) %>% 
  relocate(split, .after = mevent_nyear)

# OBS: could further make some historical aggregates

# Could transform diastolic and systolic to a categorical variable using
# bp = case_when(diastolic <  80 & systolic < 120 ~ 0,
#                diastolic <  80 & systolic < 129 ~ 1,
#                diastolic <  89 | systolic < 139 ~ 2,
#                diastolic < 120 | systolic < 180 ~ 3,
#                T ~ 4)

# find summaries for all predictors (including the new summary variables)
fin_sum <- 
  final_data %>%
  filter(split == "train") %>% 
  select(-c(id, date, split)) %>% 
  summarise(across(everything(), ~ min(., na.rm = T), 
                   .names = "min_{.col}"), 
            across(-starts_with("min_"), ~ max(., na.rm = T), 
                   .names = "max_{.col}"))

# Normalize predictors
final_data <- final_data %>% 
  mutate(across(-c(id, split, date),
                ~ (. - fin_sum[[paste0("min_",  cur_column())]]) /
                  (fin_sum[[paste0("max_", cur_column())]] - fin_sum[[paste0("min_",  cur_column())]])))



# Note: as we have normalized using from the training set, the above
# scaling will correspond to using the min and max from the training set to scale.



## -----------------------------------------------------------------------------
## Data prepared for workshop 2
## -----------------------------------------------------------------------------
write.csv(final_data, "Data_ready_for_workshop2.csv", row.names = F)

File: Code/exercise_solutions/solutions/Workshop_1_data_management.py
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)


# ==============================================================================
# 1. Load data
# ==============================================================================



baseline = pd.read_csv("../../../exercise_data/raw_data/baseline_data.csv")
diag = pd.read_csv("../../../exercise_data/raw_data/diag_data.csv").drop(columns=["Unnamed: 0"], errors="ignore")
dict_ = pd.read_csv("../../../exercise_data/raw_data/dict_data.csv").drop(columns=["Unnamed: 0"], errors="ignore")
quest = pd.read_csv("../../../exercise_data/raw_data/quest_data.csv").drop(columns=["Unnamed: 0"], errors="ignore")
blood = pd.read_csv("../../../exercise_data/raw_data/blood_data.csv").drop(columns=["Unnamed: 0"], errors="ignore")
treat = pd.read_csv("../../../exercise_data/raw_data/treat_data.csv").drop(columns=["Unnamed: 0"], errors="ignore")
events = pd.read_csv("../../../exercise_data/raw_data/events.csv").drop(columns=["Unnamed: 0"], errors="ignore")
visits = pd.read_csv("../../../exercise_data/raw_data/visit_date.csv")

# ==============================================================================
# Data cleaning: rename columns
# ==============================================================================
blood = blood.rename(columns={"..record.id": "id"})
treat = treat.rename(columns={"record_id": "id"})
quest = quest.rename(columns={"date.x": "date"})

# ==============================================================================
# 2. Translate diagnosis codes
# ==============================================================================
diag = diag.merge(dict_, on="code", how="left").drop(columns=["code"])

# ==============================================================================
# 3. Join feature datasets
# ==============================================================================
# Rename date columns to 'date' for consistent merging
blood = blood.rename(columns={'sample_date': 'date'})
diag = diag.rename(columns={'sample_date': 'date'})
treat = treat.rename(columns={'treat_start_date': 'date'})
visits = visits.rename(columns={'visit_date': 'date'})

# Add visit indicator
visits["visit"] = True

# Merge all feature datasets
feat_merged = blood.merge(diag, on=['id', 'date'], how='outer') \
                 .merge(quest, on=['id', 'date'], how='outer') \
                 .merge(treat, on=['id', 'date'], how='outer') \
                 .merge(visits, on=['id', 'date'], how='outer')

# ==============================================================================
# 4. Add baseline information
# ==============================================================================
full_dat = feat_merged.merge(baseline, on='id', how='outer')

# ==============================================================================
# 5. Train-validation split (70/10/10/10)
# ==============================================================================
np.random.seed(123)
splits = pd.DataFrame({
    'id': baseline['id'],
    'split': np.random.choice(
        ["train", "validation", "calibration", "test"],
        size=len(baseline),
        p=[0.7, 0.1, 0.1, 0.1]
    )
})
full_dat = full_dat.merge(splits, on='id', how='left')

# ==============================================================================
# 6. Remove predictors with >20% missing at patient-level (train only)
# ==============================================================================
train_data = full_dat[full_dat['split'] == 'train']

# For each column (except id, date, split), calculate % of patients with all NA
all_na_counts = {}
for col in train_data.columns:
    if col in ['id', 'date', 'split']:
        continue
    na_check = train_data.groupby('id')[col].apply(lambda x: x.isnull().all())
    all_na_counts[col] = na_check.mean()

# Keep columns where <20% patients have all NA, plus essential columns
cols_to_keep = ['id', 'date', 'split']
cols_to_keep.extend([c for c in all_na_counts if all_na_counts[c] < 0.2])
full_dat = full_dat[cols_to_keep]

# ==============================================================================
# 7. Calculate age at each visit
# ==============================================================================
full_dat['date'] = pd.to_datetime(full_dat['date'], errors='coerce')
full_dat['d.birth'] = pd.to_datetime(full_dat['d.birth'], errors='coerce')
full_dat['age'] = (full_dat['date'] - full_dat['d.birth']).dt.days / 365.242

# Convert integer columns to float (like R's as.double)
int_cols = full_dat.select_dtypes(include=['int64']).columns
full_dat[int_cols] = full_dat[int_cols].astype(float)

# ==============================================================================
# 8. Summaries for imputation and normalization (mean, q1, q99)
# ==============================================================================
train_data = full_dat[full_dat['split'] == 'train']
numeric_cols = train_data.select_dtypes(include=[np.number]).columns

summaries = {}
for col in numeric_cols:
    if col in ['id', 'date', 'split']:
        continue
    summaries[f'mean_{col}'] = train_data[col].mean()
    summaries[f'q1_{col}'] = train_data[col].quantile(0.01)
    summaries[f'q99_{col}'] = train_data[col].quantile(0.99)

# ==============================================================================
# 9. Feature engineering
# ==============================================================================
full_dat = full_dat.sort_values(['id', 'date'])

# Prediction time points
full_dat['visit'] = full_dat['visit'].fillna(False)

# Sex encoding (R creates both male and female)
full_dat['male'] = (full_dat['sex'] == 'male').astype(int)
full_dat['female'] = (full_dat['sex'] == 'female').astype(int)

# Smoking encoding (R creates current, former, never)
full_dat['smoker_current'] = (full_dat['smoking'] == 'current').astype(int)
full_dat['smoker_former'] = (full_dat['smoking'] == 'former').astype(int)
full_dat['smoker_never'] = (full_dat['smoking'] == 'never').astype(int)

# Diagnoses with carry-forward
for cond in ['diabetes', 'hyperlipidemia', 'hypertension']:
    full_dat[cond] = (full_dat['diag'] == cond).astype(int)
    full_dat[cond] = full_dat.groupby('id')[cond].cumsum()

# Drop original columns
full_dat = full_dat.drop(columns=['diag', 'd.birth', 'sex', 'smoking'], errors='ignore')

# Identify measurement columns (numeric, not binary encoded)
binary_encoded = ['male', 'female', 'smoker_current', 'smoker_former', 'smoker_never',
                  'diabetes', 'hyperlipidemia', 'hypertension']
measurement_cols = [c for c in numeric_cols if c not in binary_encoded + ['id', 'date', 'split', 'age']]

# Carry-forward imputation for measurement columns
if measurement_cols:
    full_dat[measurement_cols] = full_dat.groupby('id')[measurement_cols].ffill()

# Impute remaining NA with training means (for measurement columns and age)
impute_cols = measurement_cols + ['age']
for col in impute_cols:
    if f'mean_{col}' in summaries:
        full_dat[col] = full_dat[col].fillna(summaries[f'mean_{col}'])

# Cap at q1 and q99 for measurement columns and age
norm_cols = measurement_cols + ['age']
for col in norm_cols:
    if f'q1_{col}' in summaries and f'q99_{col}' in summaries:
        full_dat[col] = full_dat[col].clip(
            lower=summaries[f'q1_{col}'],
            upper=summaries[f'q99_{col}']
        )

# Min-max normalization for measurement columns and age
for col in norm_cols:
    if f'q1_{col}' in summaries and f'q99_{col}' in summaries:
        q1 = summaries[f'q1_{col}']
        q99 = summaries[f'q99_{col}']
        full_dat[col] = (full_dat[col] - q1) / (q99 - q1)

# ==============================================================================
# 10. Filter to prediction time points only
# ==============================================================================
pred_data = full_dat[full_dat['visit'] == True].copy()
pred_data = pred_data.drop(columns=['visit'])

# Relocate split column after id
cols = pred_data.columns.tolist()
if 'split' in cols:
    cols.remove('split')
    id_idx = cols.index('id')
    cols.insert(id_idx + 1, 'split')
    pred_data = pred_data[cols]

# ==============================================================================
# 11. Define outcome: major event or death within 1 year
# ==============================================================================
# Get first death, major, or minor event per patient
event_oi = events[events['type'].isin(['death', 'major', 'minor'])] \
            .sort_values('date') \
            .groupby('id') \
            .first() \
            .reset_index()[['id', 'date']] \
            .rename(columns={'date': 'd.event'})

pred_data['date'] = pd.to_datetime(pred_data['date'])
event_oi['d.event'] = pd.to_datetime(event_oi['d.event'])

final_data = pred_data.merge(event_oi, on='id', how='left')

# Vectorized calculation: event within 1 year
one_year_later = final_data['date'] + pd.DateOffset(years=1)
final_data['event_1y'] = (
    (final_data['d.event'] >= final_data['date']) &
    (final_data['d.event'] <= one_year_later)
).fillna(False).astype(int)

# Drop unnecessary columns
final_data = final_data.drop(columns=['d.event', 'event', 'type'], errors='ignore')

# Drop duplicates                                                                                                                                 ▁ document comprehensive, but without any redundacy or repeating text. Put the .md file in the same dir as the code and Call it workshop_2_R
final_data = final_data.drop_duplicates(subset=['id', 'date'], keep='first')

print(final_data)
print(final_data.describe())

# ==============================================================================
# 12. Save final data
# ==============================================================================
final_data.to_csv("Data_ready_for_workshop2_pytest.csv", index=False)
print("Data saved to Data_ready_for_workshop2_pytest.csv")
